docs/tutorials/sampling/ResamplingTutorial.py

```diff
@@ -0,0 +1,399 @@
+# - particles 1, 3 and 4 do not get resampled as their weights are :math:`\leq 1/5`.
+#
+#
+# A total of 3 particles were sampled from stage 1 (2x particle 2, 1x particle 5), hence
+# :math:`N_1 = \sum_{i=1}^{N}N^i_1 = 3`
+#
+#
+# Original CDF with integer weights removed:
+
+plot(normalised_weights, residual=True)
+
+
+# %%
+# New CDF:
+
+floors = np.floor(normalised_weights * n_particles)
+residuals = (normalised_weights - floors/n_particles)
+plot(residuals)
+
+# %%
+# The CDF of the residual weights is carried on to stage 2.
+#
+#
+# **Stage 2**
+#
+#
+# We now look to resample the remaining :math:`N_2 = N - N_1` particles during stage 2. (Reminder:
+# N is the total number of particles we want to sample, while :math:`N_1` is the number of
+# particles we have already resampled from stage 1). The residual weights from each particle are
+# carried over from stage 1. These residual weights are normalised, and used to calculate a CDF,
+# which is shown below.
+#
+#
+# Normalised residual weight CDF:
+
+normalised_residual_weights = residuals*(1/sum(residuals))
+plot(normalised_residual_weights)
+
+# %%
+# We then use either the :class:`~.MultinomialResampler`, :class:`~.StratifiedResampler`, or
+# :class:`~.SystematicResampler` to sample the remaining :math:`N_2` particles using the CDF.
+#
+#
+# Continuing with our example, we want to resample the remaining :math:`N_2 = N - N_1 = 5 - 3 = 2`
+# particles in stage 2. Here we will choose to use the :class:`~.MultinomialResampler` (we could
+# also use :class:`~.StratifiedResampler` or :class:`~.SystematicResampler`) to sample
+# the :math:`N_2 = 2` particles from the normalised residual weight CDF - more detail on how this
+# is done can be found in the Multinomial Resampler section above.
+
+u_j = np.random.rand(2)
+
+plot(normalised_weights, u_j)
+
+# %%
+#
+# This method reduces the number of particles that are sampled through the more computationally
+# expensive methods seen above. The Residual resampler also guarantees that any particle of weight
+# greater than :math:`1/N` will be represented in the set of resampled particles.
+#
+# When using the Residual resampler in Stone Soup, the Resampler requires a property
+# 'residual_resampler'. This property defines which resampler method to use for resampling the
+# residuals in stage 2.
+# The variable must be a string value from the following: 'multinomial', 'systematic',
+# 'stratified'. If no `residual_method` variable is provided, the multinomial method will be used
+# by default.
+
+# %%
+# ESS Resampler
+# ^^^^^^^^^^^^^
+#
+# The :class:`~.ESSResampler` (Effective Sample Size) is a wrapper around another resampler. It
+# performs a check at each time step to determine whether it is necessary to resample the
+# particles. Resampling is only performed at a given time step if a defined criterion is met. By
+# default, this criterion is
+# :math:`\frac{1}{\sum \exp(2 * particle\_log\_weight)}1 \leq \frac{n\_particles}{2}`
```

Comment:
I'll leave the LATEX to you, but could you change this to the more mathematical formula found on page 17 [here](https://www.stats.ox.ac.uk/~doucet/doucet_johansen_tutorialPF2011.pdf)? And reword to something like:

This criterion is for Kish's ESS, given by
[better formula]
to be less than or equal to the threshold, a `float` which is set by default to
\frac{N}{2}

Reply:
This should now be sorted